src/librender/scene_embree.inl

```diff
@@ -65,20 +73,30 @@ Scene<Float, Spectrum>::ray_intersect_cpu(const Ray3f &ray, Mask active) const {
             if (rh.ray.tfar != ray.maxt) {
                 ScopedPhase sp(ProfilerPhase::CreateSurfaceInteraction);
                 uint32_t shape_index = rh.hit.geomID;
-                uint32_t prim_index = rh.hit.primID;
+                uint32_t prim_index  = rh.hit.primID;
+                // We get level 0 because we only support one level of instancing
+                uint32_t inst_index = rh.hit.instID[0];
 
                 // Fill in basic information common to all shapes
                 si.t = rh.ray.tfar;
                 si.time = ray.time;
                 si.wavelengths = ray.wavelengths;
-                si.shape = m_shapes[shape_index];
                 si.prim_index = prim_index;
 
-                // Create the cache for the Mesh shape
-                Float cache[2] = { rh.hit.u, rh.hit.v };
-
-                // Ask shape to fill in the rest
-                si.shape->fill_surface_interaction(ray, cache, si);
+                // If the hit is not on an instance
+                if( inst_index == RTC_INVALID_GEOMETRY_ID ) {
+                    // If the hit is not on an instance
+                    si.shape = m_shapes[shape_index];
+                    // Create the cache for the Mesh shape
+                    Float cache[2] = { rh.hit.u, rh.hit.v };
+                    // Ask shape to fill in the rest
+                    si.shape->fill_surface_interaction(ray, cache, si);
+                } else {
+                    // If the hit is on an instance
+                    si.instance = m_shapes[inst_index];
```

Comment:
`si.shape` becomes undefined/`NULL` on this branch?

IIRC, In Mitsuba 1: `shape` is *always* the intersected shape, and `instance` can contain a pointer to the `ShapeGroup` if applicable.